File: sdk/samples/03_types/python/bitsets_bitmasks.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright (c) 2025-2026 naskel.com

"""
Bitsets and Bitmasks Sample - Demonstrates DDS bit types

This sample shows how to work with bit types:
- Bitmask types (Permissions: READ, WRITE, EXECUTE, DELETE)
- Bitset types (StatusFlags: priority[4], active[1], error[1], warning[1])
"""

import sys
sys.path.insert(0, '.')

from generated.Bits import Permissions, StatusFlags, Bits


def main():
    print("=== HDDS Bitsets and Bitmasks Sample ===\n")

    # Permissions bitmask
    print("--- Permissions Bitmask ---")
    print("Permission flags:")
    print(f"  READ    = 0x{Permissions.READ:02X} ({Permissions.READ})")
    print(f"  WRITE   = 0x{Permissions.WRITE:02X} ({Permissions.WRITE})")
    print(f"  EXECUTE = 0x{Permissions.EXECUTE:02X} ({Permissions.EXECUTE})")
    print(f"  DELETE  = 0x{Permissions.DELETE:02X} ({Permissions.DELETE})")

    # Create permissions with multiple flags
    perms = Permissions(Permissions.READ | Permissions.WRITE)

    print(f"\nPermissions with READ | WRITE:")
    print(f"  bits: 0x{int(perms):02X}")
    print(f"  has READ:    {bool(perms & Permissions.READ)}")
    print(f"  has WRITE:   {bool(perms & Permissions.WRITE)}")
    print(f"  has EXECUTE: {bool(perms & Permissions.EXECUTE)}")
    print(f"  has DELETE:  {bool(perms & Permissions.DELETE)}")

    # StatusFlags bitset (dataclass with bit fields)
    print("\n--- StatusFlags Bitset ---")
    print("Bitset fields: priority[4 bits], active[1], error[1], warning[1]")

    status = StatusFlags()
    status.priority = 5
    status.active = 1
    status.warning = 1

    print(f"\nStatusFlags with priority=5, active=1, warning=1:")
    print(f"  raw bits: 0x{status.bits:02X}")
    print(f"  priority: {status.priority}")
    print(f"  active:   {status.active}")
    print(f"  error:    {status.error}")
    print(f"  warning:  {status.warning}")

    # Bits struct serialization
    print("\n--- Bits Serialization ---")
    demo = Bits(
        perms=Permissions(Permissions.READ | Permissions.EXECUTE),
        flags=status,
    )

    print("Original:")
    print(f"  perms: 0x{int(demo.perms):02X}")
    print(f"  flags.bits: 0x{demo.flags.bits:02X}")
    print(f"  flags.priority: {demo.flags.priority}")
    print(f"  flags.active: {demo.flags.active}")

    data = demo.encode_cdr2_le()
    print(f"Serialized size: {len(data)} bytes")
    print(f"Serialized: {data.hex().upper()}")

    deser, _ = Bits.decode_cdr2_le(data)
    print("Deserialized:")
    print(f"  perms: 0x{int(deser.perms):02X}")
    print(f"  flags.bits: 0x{deser.flags.bits:02X}")
    print(f"  flags.priority: {deser.flags.priority}")
    print(f"  flags.active: {deser.flags.active}")

    if demo == deser:
        print("[OK] Bits round-trip successful\n")

    # Test flag operations
    print("--- Flag Operations ---")

    flags = Permissions(0)
    print(f"Initial:      0x{int(flags):02X}")

    flags = flags | Permissions.READ
    print(f"After +READ:  0x{int(flags):02X}")

    flags = flags | Permissions.WRITE
    print(f"After +WRITE: 0x{int(flags):02X}")

    flags = flags ^ Permissions.EXECUTE
    print(f"After ^EXEC:  0x{int(flags):02X}")

    flags = flags & ~Permissions.READ
    print(f"After -READ:  0x{int(flags):02X}")

    # All permissions
    print("\n--- All Permissions ---")
    all_perms = Permissions(Permissions.READ | Permissions.WRITE | Permissions.EXECUTE | Permissions.DELETE)
    print(f"All permissions: 0x{int(all_perms):02X}")

    all_flags = StatusFlags()
    all_flags.priority = 15
    all_flags.active = 1
    all_flags.error = 1
    all_flags.warning = 1

    all_demo = Bits(perms=all_perms, flags=all_flags)
    all_data = all_demo.encode_cdr2_le()
    all_deser, _ = Bits.decode_cdr2_le(all_data)
    print(f"Round-trip perms: 0x{int(all_deser.perms):02X}")
    print(f"Round-trip flags: priority={all_deser.flags.priority}, active={all_deser.flags.active}, "
          f"error={all_deser.flags.error}, warning={all_deser.flags.warning}")

    print("\n=== Sample Complete ===")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: sdk/samples/03_types/cpp/bitsets_bitmasks.cpp
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright (c) 2025-2026 naskel.com

/**
 * Bitsets and Bitmasks Sample - Demonstrates DDS bit types
 *
 * This sample shows how to work with bit types:
 * - Bitmask: Permissions (READ, WRITE, EXECUTE, DELETE)
 * - Bitset: StatusFlags (priority:4, active:1, error:1, warning:1)
 * - Bits struct wrapping both
 */

#include <iostream>
#include <iomanip>
#include <cstdint>
#include "generated/Bits.hpp"

using namespace hdds_samples;

int main() {
    std::cout << "=== HDDS Bitsets and Bitmasks Sample ===\n\n";

    // Permissions bitmask
    std::cout << "--- Permissions Bitmask ---\n";
    std::cout << "Permission flags:\n";
    std::cout << std::hex << std::uppercase;
    std::cout << "  READ    = 0x" << std::setw(2) << std::setfill('0')
              << static_cast<uint64_t>(Permissions::READ) << std::dec
              << " (" << static_cast<uint64_t>(Permissions::READ) << ")\n";
    std::cout << std::hex;
    std::cout << "  WRITE   = 0x" << std::setw(2)
              << static_cast<uint64_t>(Permissions::WRITE) << std::dec
              << " (" << static_cast<uint64_t>(Permissions::WRITE) << ")\n";
    std::cout << std::hex;
    std::cout << "  EXECUTE = 0x" << std::setw(2)
              << static_cast<uint64_t>(Permissions::EXECUTE) << std::dec
              << " (" << static_cast<uint64_t>(Permissions::EXECUTE) << ")\n";
    std::cout << std::hex;
    std::cout << "  DELETE  = 0x" << std::setw(2)
              << static_cast<uint64_t>(Permissions::DELETE) << std::dec
              << " (" << static_cast<uint64_t>(Permissions::DELETE) << ")\n";

    // Create permissions with multiple flags
    Permissions perms = Permissions::READ | Permissions::WRITE;

    std::cout << "\nPermissions with READ | WRITE:\n";
    std::cout << std::hex;
    std::cout << "  bits: 0x" << std::setw(2) << static_cast<uint64_t>(perms) << std::dec << "\n";
    std::cout << "  has READ:    " << std::boolalpha
              << (static_cast<uint64_t>(perms & Permissions::READ) != 0) << "\n";
    std::cout << "  has WRITE:   "
              << (static_cast<uint64_t>(perms & Permissions::WRITE) != 0) << "\n";
    std::cout << "  has EXECUTE: "
              << (static_cast<uint64_t>(perms & Permissions::EXECUTE) != 0) << "\n";
    std::cout << "  has DELETE:  "
              << (static_cast<uint64_t>(perms & Permissions::DELETE) != 0) << "\n";

    // StatusFlags bitset
    std::cout << "\n--- StatusFlags Bitset ---\n";
    StatusFlags flags{};
    flags.set_priority(5);
    flags.set_active(1);
    flags.set_error(0);
    flags.set_warning(1);

    std::cout << "StatusFlags:\n";
    std::cout << "  priority: " << flags.get_priority() << "\n";
    std::cout << "  active:   " << flags.get_active() << "\n";
    std::cout << "  error:    " << flags.get_error() << "\n";
    std::cout << "  warning:  " << flags.get_warning() << "\n";
    std::cout << std::hex;
    std::cout << "  packed:   0x" << std::setw(4) << flags.to_uint64() << std::dec << "\n";

    // Bits struct serialization
    std::cout << "\n--- Bits Serialization ---\n";
    Bits demo;
    demo.perms = Permissions::READ | Permissions::EXECUTE;
    demo.flags = StatusFlags{};
    demo.flags.set_priority(3);
    demo.flags.set_active(1);
    demo.flags.set_warning(0);

    std::cout << "Original:\n";
    std::cout << std::hex;
    std::cout << "  permissions: 0x" << std::setw(2) << static_cast<uint64_t>(demo.perms) << "\n";
    std::cout << "  flags packed: 0x" << std::setw(4) << demo.flags.to_uint64() << "\n";
    std::cout << std::dec;

    std::uint8_t buf[4096];
    int len = demo.encode_cdr2_le(buf, sizeof(buf));
    std::cout << "Serialized size: " << len << " bytes\n";
    std::cout << "Serialized: ";
    for (int i = 0; i < len; ++i) {
        std::cout << std::hex << std::setw(2) << std::setfill('0')
                  << static_cast<int>(buf[i]);
    }
    std::cout << std::dec << "\n";

    Bits deser;
    deser.decode_cdr2_le(buf, (std::size_t)len);
    std::cout << "Deserialized:\n";
    std::cout << std::hex;
    std::cout << "  permissions: 0x" << std::setw(2) << static_cast<uint64_t>(deser.perms) << "\n";
    std::cout << "  flags packed: 0x" << std::setw(4) << deser.flags.to_uint64() << "\n";
    std::cout << std::dec;

    if (static_cast<uint64_t>(demo.perms) == static_cast<uint64_t>(deser.perms) &&
        demo.flags.to_uint64() == deser.flags.to_uint64()) {
        std::cout << "[OK] Bits round-trip successful\n\n";
    }

    // Test flag operations
    std::cout << "--- Flag Operations ---\n";
    Permissions p = static_cast<Permissions>(0);
    std::cout << std::hex;
    std::cout << "Initial:      0x" << std::setw(2) << static_cast<uint64_t>(p) << "\n";

    p = p | Permissions::READ;
    std::cout << "After +READ:  0x" << std::setw(2) << static_cast<uint64_t>(p) << "\n";

    p = p | Permissions::WRITE;
    std::cout << "After +WRITE: 0x" << std::setw(2) << static_cast<uint64_t>(p) << "\n";

    p = p ^ Permissions::EXECUTE;
    std::cout << "After ^EXEC:  0x" << std::setw(2) << static_cast<uint64_t>(p) << "\n";

    p = p & ~Permissions::READ;
    std::cout << "After -READ:  0x" << std::setw(2) << static_cast<uint64_t>(p) << "\n";
    std::cout << std::dec;

    // All permissions
    std::cout << "\n--- All Permissions ---\n";
    Permissions all_perms = Permissions::READ | Permissions::WRITE |
                            Permissions::EXECUTE | Permissions::DELETE;
    std::cout << std::hex;
    std::cout << "All permissions: 0x" << std::setw(2) << static_cast<uint64_t>(all_perms) << "\n";

    Bits all_demo;
    all_demo.perms = all_perms;
    all_demo.flags = StatusFlags{};
    std::uint8_t abuf[4096];
    int alen = all_demo.encode_cdr2_le(abuf, sizeof(abuf));
    Bits all_deser;
    all_deser.decode_cdr2_le(abuf, (std::size_t)alen);
    std::cout << "Round-trip:      0x" << std::setw(2) << static_cast<uint64_t>(all_deser.perms) << "\n";
    std::cout << std::dec;

    std::cout << "\n=== Sample Complete ===\n";
    return 0;
}
